cmd/zrepl/mgmt_conn.c

```diff
@@ -538,6 +560,115 @@ uzfs_zvol_rebuild_status(uzfs_mgmt_conn_t *conn, zvol_io_hdr_t *hdrp,
 	return (reply_data(conn, &hdr, &status_ack, sizeof (status_ack)));
 }
 
+/*
+ * Iterate through all finished async tasks and send replies to clients.
+ */
+static int
+finish_async_tasks(void)
+{
+	uzfs_mgmt_conn_t *conn;
+	async_arg_t *async_task;
+	int rc;
+
+	mutex_enter(&async_tasks_mtx);
+	while ((async_task = SLIST_FIRST(&async_tasks)) != NULL) {
+		SLIST_REMOVE_HEAD(&async_tasks, task_next);
+		/* connection could have been removed in the meantime */
+		mutex_enter(&conn_list_mtx);
+		SLIST_FOREACH(conn, &uzfs_mgmt_conns, conn_next) {
+			if (conn == async_task->ctx)
+				break;
+		}
+		mutex_exit(&conn_list_mtx);
+		if (conn != NULL) {
+			rc = reply_nodata(conn, async_task->status,
+			    async_task->hdr.opcode, async_task->hdr.io_seq);
+		}
+		kmem_free(async_task, sizeof (*async_task));
+		if (rc != 0)
+			return (rc);
+	}
+	mutex_exit(&async_tasks_mtx);
+	return (0);
+}
+
+/*
+ * Perform the command (in async context).
+ *
+ * Currently we have only snapshot commands which are async. We might need to
+ * make the code & structures more generic if we add more commands.
+ */
+static void
+uzfs_zvol_do_async_command(void *arg)
+{
+	async_arg_t *async_task = arg;
+	zvol_info_t *zinfo = async_task->zinfo;
+	char *snapname = async_task->payload;
+	char *dataset;
+	int rc;
+
+	switch (async_task->hdr.opcode) {
+	case ZVOL_OPCODE_SNAP_CREATE:
+		rc = dmu_objset_snapshot_one(zinfo->name, snapname);
+		if (rc != 0) {
+			fprintf(stderr, "Snapshot create failed: %d\n", rc);
+			async_task->status = ZVOL_OP_STATUS_FAILED;
+		} else {
+			async_task->status = ZVOL_OP_STATUS_OK;
+		}
+		break;
+	case ZVOL_OPCODE_SNAP_DESTROY:
+		dataset = kmem_asprintf("%s@%s", zinfo->name, snapname);
+		rc = dsl_destroy_snapshot(dataset, B_FALSE);
+		strfree(dataset);
+		if (rc != 0) {
+			fprintf(stderr, "Snapshot destroy failed: %d\n", rc);
+			async_task->status = ZVOL_OP_STATUS_FAILED;
+		} else {
+			async_task->status = ZVOL_OP_STATUS_OK;
+		}
+		break;
+	default:
+		ASSERT(0);
+	}
+
+	/* Drop the async op if event loop thread has terminated */
+	mutex_enter(&async_tasks_mtx);
+	if (mgmt_eventfd >= 0) {
+		uint64_t val = 1;
+
+		SLIST_INSERT_HEAD(&async_tasks, async_task, task_next);
```

Comment:
We insert at head, would it be a problem because by doing so we are sending ack in reverse time order i.e. cmd which come first will be acknowledged at last. No ?